_posts/2024-06-14-granite-rag.adoc

```diff
@@ -0,0 +1,286 @@
+
+The secret of an effective RAG implementation lies in how the text segments are computed.
+In our application, we will follow a straightforward approach, but more advanced techniques are available and often required.
+
+The embedding model is responsible for converting text into a vector representation.
+For simplicity, we use an in-process embedding model (`BGE-small`).
+Other options, like the Universal Angle Embedding, are available, but we'll stick with BGE-small for simplicity.
+
+=== Retrieval
+
+The second step is the retrieval process.
+When a user inputs a query, the application searches the vector database to find the most relevant text segments.
+
+To achieve this, the application computes the vector representation of the user query using the **same** embedding model and compares it with the vector representations of the stored text segments in Infinispan.
+It selects the most relevant text segments based on the similarity between the query vector and the text segment vectors.
+
+Then, it augments the user query with the retrieved text segments and sends it to the AI Service.
+Note that until this step, the LLM is not used.
+
+== Implementing the chatbot
+
+Enough theory—let's dive into the implementation.
+You can find the final version on https://github.com/cescoffier/quarkus-granite-rag-demo[GitHub].
+
+=== Used extensions and dependencies
+
+To implement our chatbot, we rely on the following Quarkus extensions:
+
+* `quarkus-langchain4j-openai`: Integrates LLM providers using the OpenAI API, suitable for both Instructlab and Podman AI Studio.
+* `quarkus-websockets-next`: Provides support for WebSocket communication.
+* `quarkus-langchain4j-infinispan`: Integrates Infinispan with Langchain4J, allowing us to store and retrieve vector representations of text segments.
+* `quarkus-web-bundler`: Bundles frontend resources with the Quarkus application.
+
+We also need a specific dependency to use the BGE-small embedding model:
+
+[source,xml]
+----
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-embeddings-bge-small-en-q</artifactId>
+    <version>0.31.0</version>
+</dependency>
+----
+
+=== Configuration
+
+We need a bit of configuration to ensure our application uses Granite and sets up the Infinispan database correctly:
+
+[source, properties]
+----
+# Configure the Infinispan vectors:
+quarkus.langchain4j.infinispan.dimension=384
+
+# Configure the OpenAI service to use instruct lab:
+quarkus.langchain4j.openai.base-url=http://localhost:8000/v1
+quarkus.langchain4j.openai.timeout=60s
+
+quarkus.langchain4j.embedding-model.provider=dev.langchain4j.model.embedding.BgeSmallEnQuantizedEmbeddingModel
+----
+
+The first line configures the dimension of the vectors stored in Infinispan, which depends on the embedding model (BGE-small in our case).
+The second and third lines configure the OpenAI service to use Instructlab.
+You can replace the base URL with the one for Podman AI Studio if you prefer.
```

Comment:
I think we should explain here that these services implement the JSON API that OpenAI SaaS defines and that is why we can do this